src/System.Linq/src/System/Linq/Set.cs

```diff
@@ -191,6 +191,27 @@ internal List<TElement> ToList()
         }
 
         /// <summary>
+        /// Creates a hashset from the items in this set.
+        /// </summary>
+        /// <returns>A list of the items in this set.</returns>
+        internal HashSet<TElement> ToHashSet(IEqualityComparer<TElement> comparer)
+        {
+#if DEBUG
+            Debug.Assert(!_haveRemoved, "Optimised ToHashSet cannot be called if Remove has been called.");
+#endif
+            int count = _count;
+
+            HashSet<TElement> hashSet = new HashSet<TElement>(count, comparer);
```

Comment:
Maybe we can do that comparison further up the stack, and just skip going through `Set` in that case, leaving this as a case where we know `count` definitely shouldn't be used.